### PR TITLE
feat(extension-callout): add configurable emoji

### LINKED ISSUE #871: Add configurable emoji to callouts
### Motivation

Today, callout nodes have a semantic marker: info, warning, error, success. Depending on the type, the callout can be styled. For example, an error callout could have a red background.

We found the 4 semantic categories too limiting for many use cases - especially "info" callouts could benefit from more, context-specific sub categories (see example below).

### Proposed solution

Users should be allowed to configure an emoji/icon for the callout freely, i.e. independent of the semantic marker.

### Example use cases

In these examples, the emoji/icon conveys better the message than a simple semantic callout:
image

Note: The emoji/icon allows users to create error/warning/success callouts (similar to semantic callouts):
image

### Inspirations

In Notion.so, clicking the emoji of the callout allows users to select a custom emoji:
image

Callout Ui design comes with icons for callouts: https://dribbble.com/shots/12882818-Callout-Ui-design-Figma-React-UI-kit

### PR DESCRIPTION
**Pull request by Gaaaga**
### Description

Added configurable emoji to callouts. 

Added a new type: `blank`, so users are able to use a neutral background when they use the emoji as a semantic marker.

Fixes #871

### Checklist

- [x] I have read the [**contributing**](https://github.com/remirror/remirror/blob/HEAD/docs/contributing.md) document.
- [x] My code follows the code style of this project and `pnpm fix` completed successfully.
- [x] I have updated the documentation where necessary.
- [x] New code is unit tested and all current tests pass when running `pnpm test`.

### Screenshots

https://user-images.githubusercontent.com/16329365/123252988-7f37d500-d51f-11eb-9c7c-90eea5e607d2.mov

https://user-images.githubusercontent.com/24715727/123256602-bc05cb00-d523-11eb-8989-e11af894c3c4.mov

I truly appreciate @ocavue 's patient guidance, helping me dive right in the remirror project and have a better understanding of how it works.

### CODE REVIEW

File: packages/remirror__extension-callout/src/callout-extension.ts [hunk 1]
```diff
@@ -46,6 +51,33 @@ export class CalloutExtension extends NodeExtension<CalloutOptions> {
...
+   */
+  createNodeViews(): NodeViewMethod {
+    return (node: ProsemirrorNode, view: EditorView, getPos: boolean | (() => number)) => {
+      const { type } = node.attrs;
+      const pos = getPos as () => number;
+      const dom = document.createElement('div');
+      const contentDOM = document.createElement('div');
+      const emojiDOM = document.createElement('div');
+      const { renderEmoji } = this.options;
+      const emoji = renderEmoji(node, view, pos);
```
Comment (ocavue) [inline packages/remirror__extension-callout/src/callout-extension.ts:66]:
```suggestion
      const dom = document.createElement('div');
      const contentDOM = document.createElement('div');
      const emojiDOM = document.createElement('div');
      const { renderEmoji } = this.options;
      const emoji = renderEmoji(node, view, getPos as () => number);
```

File: packages/remirror__extension-callout/src/callout-extension.ts [hunk 2]
```diff
@@ -56,6 +88,7 @@ export class CalloutExtension extends NodeExtension<CalloutOptions> {
       attrs: {
         ...extra.defaults(),
         type: { default: defaultType },
+        emoji: { default: '💡' },
```
Comment (ocavue) [inline packages/remirror__extension-callout/src/callout-extension.ts:91]: Maybe it's better to allow users to change the default emoji with extension options, like `defaultType` above.

File: packages/remirror__extension-callout/src/callout-extension.ts [hunk 3]
```diff
@@ -74,9 +107,8 @@ export class CalloutExtension extends NodeExtension<CalloutOptions> {
         ...(override.parseDOM ?? []),
       ],
       toDOM: (node) => {
-        const { type, ...rest } = omitExtraAttributes(node.attrs, extra);
+        const { type, emoji, ...rest } = omitExtraAttributes(node.attrs, extra);
```
Comment (ocavue) [inline packages/remirror__extension-callout/src/callout-extension.ts:110]: This variable is not used here

File: packages/remirror__extension-callout/src/callout-extension.ts [hunk 4]
```diff
@@ -15,12 +15,16 @@ import {
...
+  NodeViewMethod,
   omitExtraAttributes,
+  ProsemirrorNode,
   toggleWrap,
 } from '@remirror/core';
 import { Fragment, Slice } from '@remirror/pm/model';
 import { TextSelection } from '@remirror/pm/state';
+import { EditorView } from '@remirror/pm/view';
 
+import { ExtensionCalloutTheme } from '../../remirror__core/node_modules/@remirror/theme/src';
```
Comment (ocavue) [inline packages/remirror__extension-callout/src/callout-extension.ts:27]:
```suggestion
import { ExtensionCalloutTheme } from '@remirror/theme';
```

You may need to add `@remirror/theme` to `packages/remirror__extension-callout/package.json`.

**Comment by ocavue** [comment]
The copy-and-paste of emoji doesn't work. In the video below, the background was copied but the emoji wasn't.


https://user-images.githubusercontent.com/24715727/123033083-ba4fe080-d419-11eb-8da3-98a8c6ac4931.mov


We could do it by adding a new data attribute to the DOM. Something like below should work.

```diff
      parseDOM: [
        {
          tag: `div[${dataAttributeType}]`,
          getAttrs: (node) => {
            if (!isElementDomNode(node)) {
              return false;
            }

            const rawType = node.getAttribute(dataAttributeType);
            const type = getCalloutType(rawType, validTypes, defaultType);
+           const emoji = node.getAttribute(dataAttributeEmoji);
            const content = node.textContent;
-           return { ...extra.parse(node), type, content };
+           return { ...extra.parse(node), type, emoji, content };
          },
        },
        ...(override.parseDOM ?? []),
      ],
      toDOM: (node) => {
        const { type, emoji, ...rest } = omitExtraAttributes(node.attrs, extra);
-       const attributes = { ...extra.dom(node), ...rest, [dataAttributeType]: type };
+       const attributes = { ...extra.dom(node), ...rest, [dataAttributeType]: type, [dataAttributeEmoji]: emoji };
        return ['div', attributes, 0];
      },
```

**Comment by Gaaaga** [comment]
@ocavue   Thanks for your review, solved.

**Review by ocavue — APPROVED** [review]

**Comment by ocavue** [comment]
@whawker Do you have any time to review this PR. I have checked the code but would like to hear your opinion. You can find a complete example in the storybook.

File: .changeset/heavy-fishes-hammer.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+'@remirror/extension-callout': minor
+'@remirror/styles': minor
+'@remirror/theme': minor
+---
+
+Added configurable emoji at the front of callout extension. Added a new type 'blank' to callout extension.
```
Comment (ifiokjr) [inline .changeset/heavy-fishes-hammer.md:7]:
```suggestion
- Added configurable emoji to the start of the `CalloutExtension`. 
- Added a new type 'blank' to the `CalloutExtension`.
```

**Review by ifiokjr — CHANGES_REQUESTED** [review]
Great stuff 👍 

Just one minor change to make in the changeset, and it'll be ready to merge.

**Comment by Gaaaga** [comment]
Thanks for the review, @ifiokjr !

**Review by ifiokjr — APPROVED** [review]
I'm happy to merge this now.

**Comment by whawker** [comment]
@Gaaaga thanks for this!! Apologies for the very late reply, I've been a bit snowed under with my own work.

The code looks good, just one thing seems missing to me - I feel the callout extension should expose a command to change the emoji, rather than expecting the user to use `setNodeMarkup`.

```ts
const transaction = view.state.tr.setNodeMarkup(getPos(), undefined, {
  emoji: selection.emoji,
  type,
});
view.dispatch(transaction);
```

In my mind the expected behaviour is to import Remirror's `useCommands` hook and utilise a dedicated command.

```ts
const { toggleCalloutEmoji } = useCommands(); 
```

As a reference, I've created [this CodeSandbox](https://codesandbox.io/s/remirror-pr-000-pr70635-forked-3dss6?file=/src/Emojis.tsx) which uses `customDispatch` from `useCommands`, but ideally it would use an exposed command like above.

**Comment by Gaaaga** [comment]
Thanks @whawker ,  very constructive advice! I'm gonna take your advice and open another PR to improve it :)

**Pull request merged**